File: AI/Assignment 1/DFS.java
import java.util.*;
import java.io.*;

class Node{
    String state;
    List<String> path;
    Node(String state, List<String> path){
        this.state = state;
        this.path = path;
        this.path.add(state);
    }
}

public class DFS{
    //goal state
    static final String goal = "123456780";
    //moves
    static int[][] moves = {
        {1, 0}, {-1, 0}, {0, 1}, {0, -1}
    };

    static final int MAX_DEPTH = 20;
    public static void dfs(String start){
        Set<String> visited = new HashSet<>();
        boolean found = dfsRecursive(new Node(start, new ArrayList<>()), visited, 0);
        if(!found){
            System.out.println("No solution found");
        }
    }
    private static boolean dfsRecursive(Node current, Set<String> visited, int depth){
        if(depth > MAX_DEPTH) return false;

        visited.add(current.state);
        if (current.state.equals(goal)) {
            System.out.println("Goal reached at depth " + depth + "!");
            printPath(current.path);
            return true;
        }
        int zeroIndex = current.state.indexOf('0');
        int x = zeroIndex / 3;
        int y = zeroIndex % 3;

        //exploring in DFS irder
        for(int[] move : moves){
            int newX = x + move[0];
            int newY = y + move[1];
            
            if (newX >= 0 && newX < 3 && newY >= 0 && newY < 3) {

                String newState = swap(current.state, zeroIndexm newX * 3 + newY);
                if(!visited.contains(newState)){
                    Node nextNode = new Node(newState, current.path);
                    if(dfsRecursive(nextNode, visited, depth + 1)){
                        return true;
                    }
       }
        }
    }
    return false;

}

//swapping two character in state string
public static String swap(String state, int i, int j){
    char[] arr = state.toCharArray();
    char temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
    return new String(arr);
}

private static void printState(String state){
    for (int i = 0; i < 9; i++) {
            if (i % 3 == 0) System.out.println();
            System.out.print(state.charAt(i) + " ");
        }
        System.out.println("\n---------");
}

private static void printPath(List<String> path) {
        int step = 0;
        for (String state : path) {
            System.out.println("Step " + step + ":");
            printState(state);
            step++;
        }
    }   

public static void main(String[] args) {
    String start = "123405678";
    System.out.println("Starting DFS...");
    printState(start);
    dfs(start);
}

File: AI/Assignment1/DFS.cpp
#include <iostream>
#include <set>
#include <map>
#include <vector>
#include <algorithm>

using namespace std;

string goal = "123456780"; // goal state

// Print 3x3 puzzle
void printState(string s)
{
    for (int i = 0; i < 9; i++)
    {
        if (i % 3 == 0)
            cout << "\n";
        cout << s[i] << " ";
    }
    cout << "\n";
}

vector<int> moves = {-3, 3, -1, 1};

// Check if move is valid
bool isValid(int pos, int move)
{
    if (pos % 3 == 0 && move == -1)
        return false;
    if (pos % 3 == 2 && move == 1)
        return false;
    int newPos = pos + move;
    return (newPos >= 0 && newPos < 9);
}

bool DFSUtil(string state, set<string> &visited, map<string, string> &parent, int depth, int limit)
{
    if (state == goal)
        return true;
    if (depth >= limit)
        return false;

    visited.insert(state);
    int pos = state.find('0');

    for (int m : moves)
    {
        if (isValid(pos, m))
        {
            string next = state;
            swap(next[pos], next[pos + m]);
            if (!visited.count(next))
            {
                parent[next] = state;
                if (DFSUtil(next, visited, parent, depth + 1, limit))
                    return true;
            }
        }
    }
    return false;
}

void DFS(string start, int limit = 20)
{
    set<string> visited;
    map<string, string> parent;

    if (DFSUtil(start, visited, parent, 0, limit))
    {
        cout << "\nDFS Solution Path:\n";
        vector<string> path;
        string state = goal;
        while (state != "")
        {
            path.push_back(state);
            state = parent[state];
        }
        reverse(path.begin(), path.end());
        for (auto &s : path)
        {
            printState(s);
            cout << "-----\n";
        }
    }
    else
    {
        cout << "No solution found with DFS (limit = " << limit << ").\n";
    }
}

int main()
{
    string start = "103425786"; // Change start state
    cout << "Start State:";
    printState(start);
    DFS(start, 20);
    return 0;
}
